compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.cpp

```diff
@@ -0,0 +1,332 @@
+  }
+  // We will now capture those dimensions where L2 memory was split. The way we
+  // do this is by checking all L2->L1 DmaOps' source offset and marking those
+  // dimensions which are not equal to at least one of the source offsets.
+  DenseSet<unsigned> splitDimsSetForL2;
+  SmallVector<unsigned> splitDimsForL2;
+  for (unsigned i = 1, n = l2ToL1DmaOps.size(); i < n; i++) {
+    if (l2ToL1DmaOps[i].getSourceObjectFifo() != sourceObjectFifo) {
+      l2ToL1DmaOps[i]->emitRemark() << "has different source objectfifo";
+      sourceObjectFifo->emitRemark() << "is the expected source objectfifo";
+      return failure();
+    }
+    SmallVector<OpFoldResult> sourceOffsets =
+        l2ToL1DmaOps[i].getSourceMixedOffsets();
+    for (unsigned j = 0, m = baseSourceOffsets.size(); j < m; j++) {
+      if (baseSourceOffsets[j] != sourceOffsets[j] &&
+          !splitDimsSetForL2.contains(j)) {
+        splitDimsForL2.push_back(j);
+        splitDimsSetForL2.insert(j);
+      }
+    }
+  }
+  std::sort(splitDimsForL2.begin(), splitDimsForL2.end());
+
+  if (failed(verifySplitDimensionConstraint(splitDimsForL2))) {
+    l2ToL1DmaOps[0]->emitRemark()
+        << "cannot split L2 logicalobjectfifo because of non-contiguous split "
+           "dimensions inferred";
+    return failure();
+  }
+
+  // Fetch the L3 -> L2 Dma Op corresponding to the L2 buffer as target.
+  AMDAIE::DmaCpyNdOp l3ToL2DmaOp;
+  DenseSet<Operation *> toBeErased;
+  for (Operation *objFifoUserOp : sourceObjectFifo->getUsers()) {
+    if (auto dmaOp = dyn_cast<AMDAIE::DmaCpyNdOp>(objFifoUserOp);
+        dmaOp.getTargetObjectFifo() == sourceObjectFifo) {
+      l3ToL2DmaOp = dmaOp;
+      toBeErased.insert(dmaOp);
+      break;
+    }
+  }
+  toBeErased.insert(sourceAllocOp);
+  toBeErased.insert(sourceObjectFifo);
+
+  SmallVector<OpFoldResult, 4> staticL2AsTargetOffsets =
+      l3ToL2DmaOp.getTargetMixedOffsets();
+  SmallVector<OpFoldResult, 4> staticL2AsTargetSizes =
+      l3ToL2DmaOp.getTargetMixedSizes();
+  SmallVector<OpFoldResult, 4> staticL2AsTargetStrides =
+      l3ToL2DmaOp.getTargetMixedStrides();
+  SmallVector<int64_t, 4> l2ShapeAsTarget = llvm::to_vector(
+      cast<MemRefType>(l3ToL2DmaOp.getTargetObjectFifo().getMemref().getType())
+          .getShape());
+  OpFoldResult zeroVal = getAsIndexOpFoldResult(context, 0);
+  OpFoldResult oneVal = getAsIndexOpFoldResult(context, 1);
+  // Update split dimensions' offset/size for L2 as target . We can afford to do
+  // this here because it's going to be the same for all L3->L2 splits. Here we
+  // are setting offset = 0 and size = 1.
+  for (unsigned dim : splitDimsForL2) {
+    staticL2AsTargetOffsets[dim] = zeroVal;
+    staticL2AsTargetSizes[dim] = oneVal;
+    l2ShapeAsTarget[dim] = 1;
+  }
+  SmallVector<unsigned> nonSplitDimsForL2;
+  for (unsigned dim = 0, n = staticL2AsTargetSizes.size(); dim < n; dim++) {
+    if (splitDimsSetForL2.contains(dim)) continue;
+    nonSplitDimsForL2.push_back(dim);
+  }
```

Comment:
After `verifySplitDimensionConstraint`, `nonSplitDimsForL2` is just the following I think:
```
SmallVector<size_t> nonSplitDimsForL2(staticL2AsTargetSizes.size() - splitDimsForL2.size());
std::iota(nonSplitDimsForL2.begin(), nonSplitDimsForL2.end(), splitDimsForL2.size())
```